src/coverage/coverageService.ts

```diff
@@ -0,0 +1,163 @@
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      "/Users/brian/esp-idf/components/",
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--html",
+    ],
+    dirPath
+  );
+  return result;
+}
+
+function _runCmd(cmd: string, args: string[], dirPath: string) {
+  const modifiedEnv = appendIdfAndToolsToPath();
+  return execChildProcess(
+    `${cmd} ${args.join(" ")}`,
+    dirPath,
+    OutputChannel.init(),
+    { env: modifiedEnv, cwd: dirPath }
+  ).catch((e) => {
+    const msg = e.message ? e.message : e;
+    Logger.error("Error on gcov cmd.\n" + msg, e);
+    OutputChannel.appendLine("Error building gcov cmd.\n" + msg);
+    return "";
+  });
+}
+
+export async function generateCoverageForEditors(
+  editors: vscode.TextEditor[],
+  dirPath: string
+) {
+  const gcovObj = await buildJson(dirPath);
+  const coveredEditors: textEditorWithCoverage[] = [];
+  for (let editor of editors) {
+    const filename = basename(editor.document.fileName);
+    const gcovObjFilePath = join(
+      "build",
+      "esp-idf",
+      "main",
+      "CMakeFiles",
+      "__idf_main.dir",
+      filename
+    );
+
+    for (const gcovFile of gcovObj.files) {
+      if (
+        gcovFile.file === gcovObjFilePath ||
+        gcovFile.file === editor.document.fileName
+      ) {
+        const coveredEditor: textEditorWithCoverage = {
+          editor,
+          coveredLines: [],
+          partialLines: [],
+          uncoveredLines: [],
+          countPerLines: [],
+          allLines: [],
+        };
+
+        for (let i = 0; i < editor.document.lineCount; i++) {
+          coveredEditor.allLines.push(editor.document.lineAt(i).range);
+        }
+
+        for (let covLine of gcovFile.lines) {
+          if (covLine && !covLine["gcovr/noncode"]) {
+            const lineRange = editor.document.lineAt(covLine.line_number - 1)
+              .range;
+
+            const rangeToDelIndex = coveredEditor.allLines.findIndex((r) => {
+              return r.isEqual(lineRange);
+            });
+            coveredEditor.allLines.splice(rangeToDelIndex, 1);
+
+            if (covLine.count > 1) {
+              coveredEditor.coveredLines.push(lineRange);
+            } else if (covLine.count === 1) {
+              coveredEditor.partialLines.push(lineRange);
+            } else {
+              coveredEditor.uncoveredLines.push(lineRange);
+            }
+            coveredEditor.countPerLines.push({
+              range: lineRange,
+              count: covLine.count,
+            });
+          }
+        }
+        coveredEditors.push(coveredEditor);
+      }
+    }
+  }
+  return coveredEditors;
+}
+
+export async function previewReport(dirPath: string) {
+  try {
+    const reportHtml = await buildHtml(dirPath);
+    const previewPanel = vscode.window.createWebviewPanel(
+      "gcoveragePreview",
+      "Coverage report",
+      vscode.ViewColumn.One
+    );
+    previewPanel.webview.html = reportHtml;
```

Comment:
Re running the command will launch multiple webviews (is this desired behaviours?)

Reply:
Fixed to singleton behavior.